{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl\n",
    "import os, sys\n",
    "from tqdm import tqdm\n",
    "\n",
    "# input_path = sys.argv[1]\n",
    "input_path = 'data/testing/testing_data.tsv'\n",
    "\n",
    "name, extension = os.path.splitext(input_path)\n",
    "output_path = f'{name}_scored.tsv'\n",
    "\n",
    "db_path = '/home/bcaron/Storage/prescored_hg38/full'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stream-load and annotate variants\n",
    "\n",
    "This scripts receives a vcf file as input and add the NCBoost score at the end of the INFO field.\n",
    "Please only input vcf files with bi-allelic variant representation (one variant per line).\n",
    "If you want to annotate a multiallelic vcf file (more than one variant per line), please use bcftools to convert it first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/home/bcaron/Storage/Dropbox/CBL/NCBoost/NCBoost-2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_nvar_in_file(input_path):\n",
    "    command = f\"wc -l {input_path}\"\n",
    "    proc = subprocess.Popen([command], stdin=subprocess.PIPE, stdout=subprocess.PIPE, shell = True)\n",
    "    out, _ = proc.communicate()\n",
    "    l_var = int(out.decode(\"utf-8\").split(' ')[0])\n",
    "    return(l_var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Chromosome: 100%|██████████| 24/24 [05:11<00:00, 12.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A total of 7554 among 7554 variants were succesfully annotated. Results have been saved in data/testing/testing_data_scored.tsv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "from src.ncboost_functions import get_chr_list, ncboost_query_score\n",
    "import tabix\n",
    "\n",
    "q_chr = 0\n",
    "n_annotated = 0\n",
    "n_var = 0\n",
    "for l_chr in tqdm(get_chr_list(), desc='Chromosome', total=len(get_chr_list()), leave=True):\n",
    "    if q_chr != l_chr:\n",
    "        q_chr = l_chr\n",
    "        ncboost_path = f\"{db_path}/WG_chr{l_chr}.tsv.gz\"\n",
    "        tb = tabix.open(ncboost_path)\n",
    "    else:\n",
    "        tb = tb\n",
    "\n",
    "    df = (\n",
    "        pl.scan_csv(input_path, separator=\"\\t\",schema_overrides={'chr':str, 'pos':int})\n",
    "        .filter(pl.col(\"chr\") == l_chr)\n",
    "        .collect(streaming=True)\n",
    "    )\n",
    "\n",
    "    ncb_score_list = []\n",
    "\n",
    "    for l_var in tqdm(df.iter_rows(named=True), desc='Variants', total=df.shape[0], leave=False):\n",
    "        l_chr = l_var['chr']\n",
    "        l_pos = l_var['pos']\n",
    "        l_ref = l_var['ref']\n",
    "        l_alt = l_var['alt']\n",
    "\n",
    "        if q_chr != l_chr:\n",
    "            q_chr = l_chr\n",
    "            ncboost_path = f\"{db_path}/WG_chr{l_chr}.tsv.gz\"\n",
    "            tb = tabix.open(ncboost_path)\n",
    "        else:\n",
    "            tb = tb\n",
    "        out = []\n",
    "        ncb_score = ncboost_query_score(l_chr=l_chr, l_pos=l_pos, l_ref=l_ref, l_alt=l_alt, tb=tb)\n",
    "        ncb_score_list.append(ncb_score)\n",
    "        if ncb_score != None:\n",
    "            n_annotated = n_annotated + 1\n",
    "    df = df.with_columns(pl.Series('NCBoost_chr_rank_percentile', ncb_score_list))\n",
    "    n_var = n_var + df.shape[0]\n",
    "    if l_chr == '1':\n",
    "        with open(output_path, \"w\") as f:\n",
    "            df.write_csv(f,\n",
    "            separator = \"\\t\", \n",
    "            include_header = True\n",
    "            )\n",
    "    else:\n",
    "        with open(output_path, \"a\") as f:\n",
    "            df.write_csv(f,\n",
    "            separator = \"\\t\",\n",
    "            include_header = False\n",
    "            )\n",
    "\n",
    "print(f'A total of {n_annotated} among {n_var} variants were succesfully annotated. Results have been saved in {output_path}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2040433011.6666667"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(6121299036-1) / 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ncboost2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
