{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl\n",
    "import os, sys\n",
    "import vcfpy\n",
    "import tabix\n",
    "import subprocess\n",
    "from tqdm import tqdm\n",
    "\n",
    "input_path = sys.argv[1]\n",
    "input_path = 'data/test.vcf'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stream-load and annotate variants\n",
    "\n",
    "This scripts receives a vcf file as input and add the NCBoost score at the end of the INFO field.\n",
    "Please only input vcf files with bi-allelic variant representation (one variant per line).\n",
    "If you want to annotate a multiallelic vcf file (more than one variant per line), please use bcftools to convert it first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "name, extension = os.path.splitext(input_path)\n",
    "output_path = f'{name}_scored.vcf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_path = 'data/test.vcf'\n",
    "os.chdir('/home/bcaron/Storage/Dropbox/CBL/NCBoost/NCBoost-2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_nvar_in_file(input_path):\n",
    "    command = f\"wc -l {input_path}\"\n",
    "    proc = subprocess.Popen([command], stdin=subprocess.PIPE, stdout=subprocess.PIPE, shell = True)\n",
    "    out, _ = proc.communicate()\n",
    "    l_var = int(out.decode(\"utf-8\").split(' ')[0])\n",
    "    return(l_var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variants:   0%|          | 0/5616 [00:00<?, ?it/s]/home/bcaron/miniconda3/envs/ncboost2/lib/python3.10/site-packages/vcfpy/parser.py:257: CannotConvertValue: - cannot be converted to Integer, keeping as string.\n",
      "  warnings.warn(\n",
      "Variants:  99%|█████████▉| 5575/5616 [00:12<00:00, 449.05it/s] \n"
     ]
    }
   ],
   "source": [
    "from src.ncboost_functions import ncboost_query_for_vcf\n",
    "\n",
    "q_chr = '0'\n",
    "db_path = '/home/bcaron/Storage/prescored_hg38/full'\n",
    "reader = vcfpy.Reader.from_path(input_path)\n",
    "reader.header.add_info_line({'ID' : 'NCBoost', 'Type' : 'Float', 'Description' : 'NCBoost hg38 score', 'Number' : 1})\n",
    "writer = vcfpy.Writer.from_path(output_path, reader.header)\n",
    "\n",
    "n_var = get_nvar_in_file(input_path)\n",
    "n_annotated = 0\n",
    "for record in tqdm(reader, desc='Variants', total=n_var):\n",
    "    l_chr = record.CHROM\n",
    "    l_pos = record.POS\n",
    "    l_ref = record.REF\n",
    "    l_alt = str(record.ALT[0]).replace(\"')\", \"\").split(\", value='\")[1]\n",
    "\n",
    "    if q_chr != l_chr:\n",
    "        q_chr = l_chr\n",
    "        ncboost_path = f\"{db_path}/WG_chr{l_chr}.tsv.gz\"\n",
    "        tb = tabix.open(ncboost_path)\n",
    "    else:\n",
    "        tb = tb\n",
    "    out = []\n",
    "    ncb_score = ncboost_query_for_vcf(l_chr=l_chr, l_pos=l_pos, l_ref=l_ref, l_alt=l_alt, tb=tb)\n",
    "    if ncb_score != None:\n",
    "        n_annotated = n_annotated + 1\n",
    "    record.INFO['NCBoost'] = ncb_score\n",
    "    writer.write_record(record)\n",
    "\n",
    "print(f'A total of {n_annotated} among {n_var} variants were succesfully annotated. Results have been saved in {output_path}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ncboost2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
